imports/node-app/plugins/discount-codes/mutations/applyDiscountCodeToCart.js

```diff
@@ -0,0 +1,134 @@
+import SimpleSchema from "simpl-schema";
+import Random from "@reactioncommerce/random";
+import ReactionError from "@reactioncommerce/reaction-error";
+import getCart from "../util/getCart.js";
+
+const inputSchema = new SimpleSchema({
+  cartId: String,
+  discountCode: String,
+  shopId: String,
+  token: {
+    type: String,
+    optional: true
+  }
+});
+
+/**
+ * @method applyDiscountCodeToCart
+ * @summary Applies a discount code to a cart
+ * @param {Object} context - an object containing the per-request state
+ * @param {Object} input - an object of all mutation arguments that were sent by the client
+ * @param {Object} input.cartId - Cart to add discount to
+ * @param {Object} input.discountCode - Discount code to add to cart
+ * @param {String} input.shopId - Shop cart belongs to
+ * @param {String} [input.token] - Cart token, if anonymous
+ * @returns {Promise<Object>} An object with the updated cart with the applied discount
+ */
+export default async function applyDiscountCodeToCart(context, input) {
+  inputSchema.validate(input);
+
+  const { cartId, discountCode, shopId, token } = input;
+  const { collections, userHasPermission, userId } = context;
+  const { Cart, Discounts } = collections;
+
+  // TODO: figure out the correct permission check here
+  // Should it be `discounts`, or `cart`?
+  // How do we determine this check if the user is the cart owner?
+  if (!userHasPermission(["admin", "owner", "discounts"], shopId)) {
+    throw new ReactionError("access-denied", "Access Denied");
+  }
+
+  let userCount = 0;
+  let orderCount = 0;
+  let cart = await getCart(context, shopId, cartId, { cartToken: token, throwIfNotFound: false });
+
+  // If we didn't find a cart, it means it belongs to another user,
+  // not the currently logged in user.
+  // Check to make sure current user has admin permission.
+  if (!cart) {
+    cart = await Cart.findOne({ _id: cartId });
+    if (!cart) {
+      throw new ReactionError("not-found", "Cart not found");
+    }
+
+    // TODO: figure out the correct permission check here
+    // Should it be `discounts`, or `cart`?
+    if (!userHasPermission(["owner", "admin", "discounts"], shopId)) {
+      throw new ReactionError("access-denied", "Access Denied");
+    }
```

Comment:
Since `discounts` is more for defining/editing discounts, I'd create a new role `discounts/apply` for here.